### ISSUE TITLE
CHARMM Drude force field

### ISSUE THREAD

**Issue by peastman** [ParmEd/ParmEd#1080]
I'd like to use ParmEd to convert the CHARMM Drude force field to OpenMM.  It uses a lot of features not found in standard Amber force fields:

- Drude particles
- Lone pairs
- NBFIX terms
- Urey-Bradley terms
- CMAP terms

I'm not sure which of these are currently supported by ParmEd and which aren't.  What would be involved in adding support for them?

**Comment by peastman** [comment]
It also uses a harmonic form for improper torsions.

**Comment by peastman** [comment]
Can anyone offer advice on this?  I'm willing to help implement the needed features, I just don't know what needs to be done or how it should be approached.

**Comment by swails** [comment]
Sorry, I never saw the email about this.

Drude particles and lone pairs are probably not in ParmEd yet (although it has some concept of virtual sites, which may suffice).  The rest are there.

You can look at `topologyobjects.py`, which contains all of these entities.

You'd basically need to add the parameter types for those new terms and then update the CHARMM parameter file parsers to read those sections and add those terms into the parameterset dict.

**Comment by peastman** [comment]
Thanks!  I'll spend a while looking over the code, then come back with more specific questions.

The `ExtraPoint` class should mostly work for lone pairs, except that CHARMM supports more complicated methods of specifying positions (https://www.charmm.org/charmm/documentation/by-version/c40b1/params/doc/lonepair/).  For Drude particles we basically need to store a bunch of other properties for certain atoms

**Comment by peastman** [comment]
For each polar atom, the force field defines the following attributes that we'll need to store:

- Alpha, the polarizability.  The presence of this attribute is what signals to add a Drude particle.
- The thole damping factor.  This is optional, but if it's missing that just means to use the default value of 1.3.
- The atom type to use for the Drude particle.  This is optional, but if it's missing that just means to use the default type DRUD.

In addition, some atoms have anisotropic polarizability.  For those, we need to store references to three other atoms that are used to define the coordinate frame, plus two scale factors.

We could add some of this information to the `Atom` class, but it might be cleaner to put all of it in a separate `Drude` class?  Or perhaps the simple parameters should go in `Atom`, but the coordinate frame definition would go in an `AnisotropicDrude` class?

I looked at the classes for AMOEBA, since those need to store very similar information, but they look incomplete.  `MultipoleFrame` stores a set of indices rather than atoms?  And it looks like they never actually get created or associated with an atom.

Am I going to be doing anything with `Structure` objects?  Or are those only used for parametrized systems, and I'll be working only with `ParameterSet` objects?

**Comment by swails** [comment]
I think a separate `Drude` class makes sense -- inheriting from `Atom` might make sense depending on how often drude particles are treated like other particles.  IIRC, ParmEd does most things by duck-typing (rather than using `isinstance` checks), so you might be fine just making `Drude` a separate class entirely that simply inherits from `_ListItem`.

> In addition, some atoms have anisotropic polarizability. For those, we need to store references to three other atoms that are used to define the coordinate frame, plus two scale factors.

It seems like those references would need to be stored in the topology object (something analogous to `Structure`) and wouldn't be of much concern when storing parameter sets?  The scale factors seem like it would make sense for a `DrudeType` class (analogous to the `AtomType` class).  `ParameterSets` carry references to types (AtomType, BondType, ...etc.) rather than the actual topology entities themselves (Atom, Bond, ...etc.).  If you're just interested in translating parameter sets, you probably don't have to worry much about Atom.  In that case, I'd just put whatever was needed in a `DrudeType` class (the two scale factors, coordinate frame definition information, ...?).

> I looked at the classes for AMOEBA

These were basically lifted from the Amber implementation of Amoeba, and never followed through on.  They are definitely incomplete.

> Am I going to be doing anything with `Structure` objects? Or are those only used for parametrized systems, and I'll be working only with `ParameterSet` objects?

`Structure` is only for parameterized systems.  If you're just working with parameter set conversion, then `ParameterSet` (and therefore `AtomType`, `BondType`, ...) objects are all you really have to worry about.

**Comment by peastman** [comment]
All these things are specified as part of residue definitions, so they aren't based on "types" exactly.  As an example, here's the definition of ALA from the force field:

```
RESI ALA           0.000 
! 
!     | 
!  HN-N 
!     |     HB1 
!     |    / 
!  HA-CA--CB-HB2 
!     |    \ 
!     |     HB3 
!   O=C 
!     | 
! 
GROUP 
ATOM N    ND2A2   -0.382  ALPHA -1.942  THOLE 0.250 
ATOM HN   HDP1A    0.272 
ATOM CA   CD31C    0.169  ALPHA -0.960  THOLE 1.078 
ATOM HA   HDA1A   -0.017 
ATOM C    CD2O1A   0.497  ALPHA -0.675  THOLE 0.295 
ATOM O    OD2C1A   0.000  ALPHA -0.651  THOLE 0.310 
ATOM LPOA LPDO1   -0.312 
ATOM LPOB LPDO1   -0.227 
GROUP 
ATOM CB   CD33A   -0.099  ALPHA -1.804  THOLE 2.080 
ATOM HB1  HDA3A    0.033 
ATOM HB2  HDA3A    0.033 
ATOM HB3  HDA3A    0.033 
 
BOND N    CA        CA   C         C    +N        CA   HA 
BOND N    HN        C    O 
BOND CA   CB        CB   HB1       CB   HB2       CB   HB3 
BOND O    LPOA      O    LPOB 
 
IMPR C   CA  +N  O   N  -C  CA HN 
CMAP -C  N   CA  C   N  CA  C  +N 
 
LONEPAIR relative LPOA O C CA distance 0.30 angle 91.0 dihe   0.0 
LONEPAIR relative LPOB O C CA distance 0.30 angle 91.0 dihe 180.0 
ANISOTROPY O C LPOA LPOB  A11  0.82322 A22  1.14332 
```

All the heavy atoms include `ALPHA` and `THOLE` parameters, so Drude particles will be added to them.  They don't specify `TYPE`, so all the Drude particles will have the default atom type DRUD.  The `LONEPAIR` lines tell us that LPOA and LPOB are virtual sites and describe how to compute their positions based on the O, C, and CA atoms.  The `ANISOTROPY` line tells us that atom O has anisotropic polarizability, that the coordinate frame is computed based on atoms C, LPOA, and LPOB, and how to scale the polarizability along the axes of that frame.

**Comment by bradakstx** [comment]
Apparently I still get email alerts for this project - neat.

I've never contributed heavily to the ParmEd code base, but I did muck with the RTF/PSF formats in order to update the Drude and Lonepair implementations in NAMD a few years back. I am also currently interested in seeing Drude-like force fields get supported by AMBER (the Onufriev lab seems to be trying to make this happen as well).

I can attest that the Drude "standards" are a multitude of hacks to make things work. For example, the PSF standard does not actually support all features of either Drude or Lonepairs without some heuristic assumptions (e.g. a Drude particle _must_ have a name that starts with "D" and _must_ have a mass in a certain range, and _must_ come after its parent). There is also a hidden parameter in the conversion of ALPHA to charges based on a magic number (5000 I think?) and the value of Coulomb's constant). It was necessary to hardcode those numbers in order to get the same charges as CHARMM.

Let me know if I can help or ping other interested parties.

**Comment by swails** [comment]
Ah, then it also looks like you'll need to work with `ResidueTemplate` and make those aware of Drudes.

**Comment by peastman** [comment]
How about the following?  I'll add two classes in topologyobjects.py:

- `DrudeAtom` inherits from `Atom`, adding fields for alpha, Thole, and Drude type.
- `DrudeAnisotropy` contains the information for an anisotropic site: atoms defining the frame and scale factors.

I'll also add an `anisotropies` field to `ResidueTemplate` listing any anisotropic sites in the residue.

I think the existing lone pair support is already sufficient for everything in this force field.

**Comment by swails** [comment]
This was fixed by #1082 I believe.

Reopen if that's not the case.
